### PR TITLE
Feature/visualize masking

### LINKED ISSUE #2145: Initiative: Visualize masking before pretraining/finetuning
### Describe the task. Describe the task. It can be a feature, a set of experiments, documentation, etc.

As many team members requested to have this feature of previewing masked source and target before starting any pretraining or fine-tuning, I am adding a script to provide a preview of how the data would look like before submitting any job. Easy to use script with one line.

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

_No response_

### URL to the design document

_No response_

### Area

- [x] datasets, data readers, data preparation and transfer
- [ ] model
- [ ] science
- [ ] infrastructure and engineering
- [x] evaluation, export and visualization
- [ ] documentation
- [ ] performance

### PR DESCRIPTION
**Pull request by wael-mika**
## Description

Implementation of the masking visualization script

## Issue Number
Closes #2145 

Is this PR a draft? Mark it as draft.

## Checklist before asking for review

-   [X] I have performed a self-review of my code
-   [X] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [X] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/utils/visualize_masking.py [hunk 1]
```diff
@@ -0,0 +1,1133 @@
...
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+"""
+Visualize masked source/target samples using the training data pipeline.
+
+This script loads a config, builds a MultiStreamDataSampler, extracts one batch,
+and plots a single variable for source and target with masking/cropping applied.
+This script can run on a cpu or logging node without GPUs.
+Please activate your .venv before running.
```
Comment (tjhunter) [inline src/weathergen/utils/visualize_masking.py:16]: this is wrong. You should not have to do that. `uv run` should handle running without activating the environment.

File: src/weathergen/utils/visualize_masking.py [hunk 2]
```diff
@@ -0,0 +1,1133 @@
...
+
+    test_cfg = wg_config.merge_configs(training_cfg, cf.get("validation_config", {}))
+    test_cfg = wg_config.merge_configs(test_cfg, cf.get("test_config", {}))
+    test_cfg = filter_config_by_enabled(test_cfg, keys_to_filter)
+    return test_cfg, VAL
+
+
+def _to_numpy(arr):
+    if isinstance(arr, torch.Tensor):
+        return arr.detach().cpu().numpy()
```
Comment (TillHae) [inline src/weathergen/utils/visualize_masking.py:297]: Why do we have to move this to cpu?

File: src/weathergen/utils/visualize_masking.py [hunk 3]
```diff
@@ -0,0 +1,1133 @@
...
+        Boolean mask over HEALPix cells (True = visible).
+    healpix_level : int
+        HEALPix resolution level (nside = 2^level).
+
+    Returns
+    -------
+    np.ndarray
+        Boolean array of same length as lats/lons indicating visibility.
+    """
+    if len(lats) == 0:
```
Comment (TillHae) [inline src/weathergen/utils/visualize_masking.py:347]: Is len(lats) == len(lons) all the time?

File: src/weathergen/utils/visualize_masking.py [hunk 4]
```diff
@@ -0,0 +1,1133 @@
...
+
+    Returns
+    -------
+    tuple[np.ndarray, np.ndarray, np.ndarray]
+        Downsampled latitude, longitude, and value arrays.
+    """
+    if max_points is None or len(vals) <= max_points:
+        return lats, lons, vals
+    idxs = rng.choice(len(vals), size=max_points, replace=False)
+    return lats[idxs], lons[idxs], vals[idxs]
```
Comment (TillHae) [inline src/weathergen/utils/visualize_masking.py:589]: This might be dangerous to do as it generates a discontinuous array that looks continuous

File: src/weathergen/utils/visualize_masking.py [hunk 5]
```diff
@@ -0,0 +1,1133 @@
...
+
+    include_full = len(vals_full) > 0
+    ncols = 3 if include_full else 2
+
+    if vmin is None or vmax is None:
+        vals_ref = vals_full
+        if len(vals_ref) == 0:
+            vals_ref = vals_src if len(vals_src) else vals_tgt
+        vmin = np.nanpercentile(vals_ref, 2)
+        vmax = np.nanpercentile(vals_ref, 98)
```
Comment (TillHae) [inline src/weathergen/utils/visualize_masking.py:663]: It doesn't make a big difference in most cases, but I think cutting the outliers can be unwanted sometimes, where we are interested in them and want to see it in our plot. Especially when evaluating masking this can be interesting.

File: src/weathergen/utils/visualize_masking.py [hunk 6]
```diff
@@ -0,0 +1,1133 @@
...
+        if target_view == "target_values":
+            target_var_idx, target_var_name = _resolve_var_idx(
+                ds.target_channels,
+                target_var_name,
+                "target",
+                strict=parsed.target_variable is not None,
+            )
+        else:
+            target_var_idx, target_var_name = _resolve_var_idx(
+                ds.source_channels,
```
Comment (TillHae) [inline src/weathergen/utils/visualize_masking.py:893]: Why do we have to use `ds.source_channels` when we have no target_tokens in our target stream?

File: src/weathergen/utils/visualize_masking.py [hunk 7]
```diff
@@ -0,0 +1,1133 @@
...
+        )
+        tgt_label = (
+            f"Target ({target_var_name}, {target_view})\n"
+            f"{_format_mask_params(_to_dict(target_meta.params))}\n"
+            f"points={len(vals_tgt)}"
+        )
+
+        pairs_data.append(
+            {
+                "pair_idx": source_idx,
```
Comment (TillHae) [inline src/weathergen/utils/visualize_masking.py:970]: Shouldn't the key "pair_idx" point to the value of `pair_indices` defined in l.829

**Review by TillHae — COMMENTED** [review]
Looks good, I just have a few concerns with your script.

**I didn't check parsing, docstrings, plotting, meta info, instance checks and didn't test the script.**

File: src/weathergen/utils/visualize_masking.py (hunk 1 above)
Comment (wael-mika) [reply src/weathergen/utils/visualize_masking.py:16]: I agree with you, but after dealing with different systems, there are different Python versions running on each of those, and I found out that using uv run is the easiest way to handle this situation. I can add a command using python as well if you agree

File: src/weathergen/utils/visualize_masking.py (hunk 2 above)
Comment (wael-mika) [reply src/weathergen/utils/visualize_masking.py:297]: This script can be run completely on cpu, we do not need GPU for it

File: src/weathergen/utils/visualize_masking.py (hunk 3 above)
Comment (wael-mika) [reply src/weathergen/utils/visualize_masking.py:347]: Yes, always. Both lats and lons are derived from the same coords array via coords[:, 0] and coords[:, 1], so they are guaranteed to have equal length at every call site.

File: src/weathergen/utils/visualize_masking.py (hunk 5 above)
Comment (wael-mika) [reply src/weathergen/utils/visualize_masking.py:663]: I agree, but this is not an evaluation plot, so extreme events are prevented here to keep the colorscale consistent.

File: src/weathergen/utils/visualize_masking.py (hunk 6 above)
Comment (wael-mika) [reply src/weathergen/utils/visualize_masking.py:893]: When visualizing masking, I am choosing only one variable, which is enough for this purpose. Source and target here represent the student and teacher view which are different and can be set up in the config

File: src/weathergen/utils/visualize_masking.py (hunk 7 above)
Comment (wael-mika) [reply src/weathergen/utils/visualize_masking.py:970]: source_idx already is a value from pair_indices (the loop iterates for source_idx in pair_indices). So "pair_idx": source_idx stores the source sample index, which is meaningful (it identifies which batch sample this is). Storing the enumeration position (0, 1, 2…) would be less informative and would make the filename pair0_, pair1_ regardless of which samples were selected.
Though the naming pair_idx vs source_idx is admittedly redundant — pair_idx could be renamed to make it clearer.

File: src/weathergen/utils/visualize_masking.py (hunk 4 above)
Comment (wael-mika) [reply src/weathergen/utils/visualize_masking.py:589]: Disagree. For scatter plots. Unlike line plots, scatter does not connect adjacent points. Each point is rendered independently, so random downsampling only reduces density. It doesn't create false continuity. The points=N label already shows how many points are displayed. 
_downsample is only used when you set up "--max-points" for high resolution data and it is not used in individual cases if not specificaly sat up

File: src/weathergen/utils/visualize_masking.py (hunk 2 above)
Comment (TillHae) [reply src/weathergen/utils/visualize_masking.py:297]: But if I have a gpu on my local machine I might want to use it for performance

**Comment by TillHae** [comment]
Thanks for clarifying. Now everything sounds logical to me.

File: src/weathergen/utils/visualize_masking.py (hunk 2 above)
Comment (wael-mika) [reply src/weathergen/utils/visualize_masking.py:297]: I believe most of users will be using HPCs as the data is stored there. That said, it is only retrieving one sample so cpu is perfectly suited even on any machine. The performance is not important here as the script is used to generate one plot before training.

**Comment by TillHae** [comment]
We should decide if this is useful enough to others to merge or too niche to add to the repo.

**Comment by clessig** [comment]
> We should decide if this is useful enough to others to merge or too niche to add to the repo.

It's useful but it should no go into the main repo but into the private one. It would also be important to have a mechansim to ensure that the code is kept functional. The masking code is relatively stable but we should know when the visualization code breaks.
